content/client-lib-development-guide/features.textile

```diff
@@ -1530,6 +1583,7 @@ h4(#token-params). TokenParams
 ** @(TK2b)@ @capability@ string - Capability requirements JSON stringified for the token. When omitted, Ably will default to the capabilities of the underlying key
 ** @(TK2c)@ @clientId@ string - A @clientId@ string to associate with this token. If @clientId@ is @null@ or omitted, then the token is prohibited from assuming a @clientId@ in any operations, however if @clientId@ is a wildcard string @'*'@, then the token is permitted to assume any @clientId@. Any other string value for @clientId@ implies that the @clientId@ is both enforced and assumed for all operations for this token
 ** @(TK2d)@ @timestamp@ long - The timestamp (in milliseconds since the epoch) of this request. Timestamps, in conjunction with the @nonce@, are used to prevent requests from being replayed. @timestamp@ is a "one-time" value, and is valid in a request, but is not validly a member of any default token params such as @ClientOptions#defaultTokenParams@
+** @(TK2e)@ @nonce@ optional string - Used to prevent against replay attacks. If not provided, then the library will automatically generate one if needed, as described by "RSA9c":#RSA9c.
```

Comment:
I'm not sure why it's in the IDL, but I wonder if the intention was that it be exposed as an internal API? For example [ably-ruby uses it in a test that repeats a nonce to trigger a specific exception](https://github.com/ably/ably-ruby/blob/65e0a24889bcc356bd0e5259f7ffa0d1133fd046/spec/acceptance/rest/client_spec.rb#L1341-L1354) (to be honest, I'd say that the `it` for that test and actual assertions it makes don't really match, but that's another matter).

It's currently exposed in cocoa, js, dotnet, ruby, and rust. It's not exposed in java or go.

I don't really have a strong opinion on this, but the general approach I've taken elsewhere was to add something (that's already in the IDL) to the spec if it's implemented in the majority of our client SDKs and if it doesn't otherwise seem _too_ controversial.

Do you have a preference for what I should do here?

Reply:
Hmm yes, I mean that Ruby test doesn't seem to make sense, since surely _not_ allowing the nonce to be set at all by the user removes the need to test that the user doesn't replay one. Also, `TokenParams#nonce` is even [documented in the API reference docs](https://docs.ably.io/api/realtime-sdk/types/#token-params) 🤔 .

So, happy for you to merge this to be consistent with what we've documented and implemented elsewhere, and I can raise the issue separately.